examples/tensorflow/summarization/run_summarization.py

```diff
@@ -0,0 +1,682 @@
+
+        It sorts the inputs according to lengths in order to minimize the padding size, with a bit of randomness for
+        the training set.
+    predict_with_generate (:obj:`bool`, `optional`, defaults to :obj:`False`):
+        Whether to use generate to calculate generative metrics (ROUGE, BLEU).
+    """
+
+    sortish_sampler: bool = field(default=False, metadata={"help": "Whether to use SortishSampler or not."})
+    predict_with_generate: bool = field(
+        default=False, metadata={"help": "Whether to use generate to calculate generative metrics (ROUGE, BLEU)."}
+    )
+
+
+# endregion
+
+# region Dataset name mappings
+summarization_name_mapping = {
+    "amazon_reviews_multi": ("review_body", "review_title"),
+    "big_patent": ("description", "abstract"),
+    "cnn_dailymail": ("article", "highlights"),
+    "orange_sum": ("text", "summary"),
+    "pn_summary": ("article", "summary"),
+    "psc": ("extract_text", "summary_text"),
+    "samsum": ("dialogue", "summary"),
+    "thaisum": ("body", "summary"),
+    "xglue": ("news_body", "news_title"),
+    "xsum": ("document", "summary"),
+    "wiki_summary": ("article", "highlights"),
+}
+# endregion
+
+
+# region Data generator
+def sample_generator(dataset, model, tokenizer, shuffle, pad_to_multiple_of=None):
+    if shuffle:
+        sample_ordering = np.random.permutation(len(dataset))
+    else:
+        sample_ordering = np.arange(len(dataset))
+    for sample_idx in sample_ordering:
+        example = dataset[int(sample_idx)]
+        # Handle dicts with proper padding and conversion to tensor.
+        example = tokenizer.pad(example, return_tensors="np", pad_to_multiple_of=pad_to_multiple_of)
+        example = {key: tf.convert_to_tensor(arr, dtype_hint=tf.int32) for key, arr in example.items()}
+        if model is not None and hasattr(model, "prepare_decoder_input_ids_from_labels"):
+            decoder_input_ids = model.prepare_decoder_input_ids_from_labels(
+                labels=tf.expand_dims(example["labels"], 0)
+            )
+            example["decoder_input_ids"] = tf.squeeze(decoder_input_ids, 0)
+        yield example, example["labels"]  # TF needs some kind of labels, even if we don't use them
+    return
+
+
+# endregion
+
+
+# region Helper functions
+def dataset_to_tf(dataset, model, tokenizer, total_batch_size, num_epochs, shuffle):
+    if dataset is None:
+        return None
+    train_generator = partial(sample_generator, dataset, model, tokenizer, shuffle=shuffle)
+    train_signature = {
+        feature: tf.TensorSpec(shape=(None,), dtype=tf.int32)
+        for feature in dataset.features
+        if feature != "special_tokens_mask"
+    }
+    if (
+        model is not None
+        and "decoder_input_ids" not in train_signature
+        and hasattr(model, "prepare_decoder_input_ids_from_labels")
+    ):
+        train_signature["decoder_input_ids"] = train_signature["labels"]
```

Comment:
I think  this should apply `model.prepare_decoder_input_ids_from_labels` to the labels, no?

Reply:
I don't think so, though I agree the code here is a bit confusing and TensorFlow-specific! I'm only setting the tensor "signatures" here (i.e. shape and dtype), not the actual tensors themselves - the conditional here checks that the model will be adding decoder_input_ids, and if so it adds a matching signature to the list of tensor signatures. Since the addition of decoder_input_ids only happens in the sample generator, I can't just check for the presence of that tensor in the original dataset.

Reply:
Oh it's just the signature, so it's fine then.